docs/proposals/ose-rebase.md

```diff
@@ -0,0 +1,107 @@
+
+Today, code in OSE lands by fetching and merging Origin manually. OSE-specific changes
+are scattered in the repo and it's hard to reason about any conflicts. This doc proposes
+moving all the changes in OSE that diverge from Origin into a set of commits that will sit
+on top of the Origin code and a Jenkins job that will be rebasing master on a daily basis.
+Rebases are going to continue to happen only for the master branch.
+
+There are two types of commits that need to be carried on top of Origin:
+1. long-term carries related to branding and release tooling (master and release branches)
+2. CVEs and other backported bugfixes (release branches only)
+
+The Origin repository is set as a remote `upstream` branch in OSE so in master we can
+simply rebase upstream/master. Before the actual rebase happens, the last commit produced
+by tito during the previous rebase should be squashed into the main tito commit (one
+of the [CARRY] commits). An intermediate commit should exist that bumps the web console
+and it should be dropped since a new updated one will be added further down the build process.
+The tag that points to the last tito commit needs to be retaged because the squash is dropping
+it. If any conflicts are hit, manual resolution will be required as it is already the case
+with `git merge`. In the future, we can extend the conflict resolution mechanism in Jenkins
+to send an e-mail to the author and committer of the carry commits.
+
+A nightly build of OSE should follow the steps below in order to rebase on top of Origin:
+```sh
+git clone git@github.com:openshift/ose.git
+pushd ose
+git remote add upstream git@github.com:openshift/origin.git
+git fetch upstream
+git checkout master
+
+# Note down the current latest tag because squashing the tito commit into specfile updates
+# will drop it and tito tag needs it further down the build process in order to determine
+# the next (new) tag.
+last_tag="$( git describe --abbrev=0 --tags )"
+
+# sanity check that ensures the last three commits are in the right order.
+#
+# The last three commits are the tito tag commit, a web console bump that got created
+# during the previous rebase, and the specfile updates:
+#
+#         .-Other [CARRY] commit
+#        /  .-[CARRY] Specfile updates
+#       /  /  .-[DROP] webconsole bump
+#      /  /  /  .-Tito tag commit
+#     /  /  /  /
+# m--c--c--w--t
+#  \           \
+#   \           `-ose/master/HEAD,v3.5.0.23
+#    `-origin/master/HEAD
+#
+sanity_check
+
+# Reset the last three commits.
+git reset HEAD~3
+
+# Squash the tito commit into the specfile updates
+git add .tito origin.spec
+git commit -m "[CARRY][BUILD] Specfile updates"
+
+# Drop the unstaged changes for the web console - a new diff will be regenerated once
+# we revendor the Web Console for OSE.
+git checkout -- pkg/assets/bindata.go pkg/assets/java/bindata.go
+
+# We need to retag, because the previous squash removed the latest tito tag.
+git tag "${last_tag}" HEAD
```

Comment:
If you need access to tags other than the latest one, you should be able to get the srpm from the mirrors they are uploaded after every build.